# Exploring Ebay Car Sales Data

### Aim of the project

_Clean the data and analyze the included used car listings._

### Dataset

The [dataset](https://www.kaggle.com/orgesleka/used-cars-database/data) of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website is used.
The dataset was originally scraped and uploaded to Kaggle.

We sampled 50,000 data points from the full dataset

In [100]:
import pandas as pd
import numpy as np

In [101]:
autos = pd.read_csv('autos.csv',encoding='Latin-1') #using encoding Latin 1

In [102]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

The above data set information suggests that most of the data is in the form of objects that is strings(15). Only a few columns have anumeric data type(5).
There are total of 20 columns with 50000 entries.

In [103]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The above cell displays the first 5 rows in the dataframe including the heading. The numbers from 0 to 50,000 make up the index of the dataframe.

### Cleaning the column names

In [104]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [105]:
change_columns =['date_crawled', 'name', 'seller','offer_type','price','abtest',
       'vehicle_type', 'registration_year','gearbox', 'power_p_s','model',
      'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'unrepaired_damage','ad_created','nr_of_pictures','postal_code',
       'last_seen']

In [106]:
autos.columns = change_columns

In [107]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  month_of_registration fuel_type  \
0    manuell        158  andere  150,000km                      3       lpg   
1  automatik        286     7er  150,000km                      6    benzin   
2    manuell        102    golf   70,000km                      7    benzin   
3  automatik         71  fortwo   70,000km                      6    benzin   
4    manuell          0   focus  150,000km                      7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Converted the column names to snakecase format so that they are easy to process and retrieve while performing further operations.

### Cleaning the dataframe

In [108]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

- To be dropped columns : seller, offer_type, abtest, gearbox, unrepaired damage
- To be cleaned : odometer
- To be converted to numeric : price(float)

Columns having mostly one value will be dropped

In [109]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace("kms","")

In [110]:
autos.rename({"odometer":"odometer_km"},inplace=True,axis=1)

In [111]:
autos["odometer_km"] = autos["odometer_km"].str.replace(",","")

In [112]:
autos["odometer_km"]=autos["odometer_km"].astype(int)

In [113]:
autos["price"] = autos["price"].str.replace(",","").str.replace("$","")
autos["price"] = autos["price"].astype(int)

In [114]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [115]:
autos["price"].describe().apply(lambda x: format(x, 'f'))

count       50000.000000
mean         9840.043760
std        481104.380500
min             0.000000
25%          1100.000000
50%          2950.000000
75%          7200.000000
max      99999999.000000
Name: price, dtype: object

In [116]:
autos[autos["price"].between(80000,9999999)].shape

(66, 20)

In [117]:
autos = autos[autos["price"].between(500,80000)]

After reading through the parameters and using multiple functions we feel the outliers are values lesser than 500 as it is almost impossible to have a car value which is less than 500. 
Also, the number of cars with the price is greater than 80,000 is just 65 and also it's greater than the mean hence they have been ignored.

In [118]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34     2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
36     2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37     2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
...                    ...                                                ...   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49967  2016-03-12 11:50:17                      VW_Passat_2_0_TDI_comfortline   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0

In [119]:
autos["date_crawled"] = autos["date_crawled"].str[:10]
autos["ad_created"] = autos["ad_created"].str[:10]
autos["last_seen"] = autos["last_seen"].str[:10]

In [120]:
autos["date_crawled"].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.025555
2016-03-06    0.014121
2016-03-07    0.036234
2016-03-08    0.033171
2016-03-09    0.032926
2016-03-10    0.032749
2016-03-11    0.033060
2016-03-12    0.037345
2016-03-13    0.015542
2016-03-14    0.036301
2016-03-15    0.033992
2016-03-16    0.029352
2016-03-17    0.031172
2016-03-18    0.012833
2016-03-19    0.034791
2016-03-20    0.038077
2016-03-21    0.037722
2016-03-22    0.032948
2016-03-23    0.032393
2016-03-24    0.028996
2016-03-25    0.031083
2016-03-26    0.032638
2016-03-27    0.031128
2016-03-28    0.034813
2016-03-29    0.033304
2016-03-30    0.033282
2016-03-31    0.031661
2016-04-01    0.033859
2016-04-02    0.035813
2016-04-03    0.038810
2016-04-04    0.036634
2016-04-05    0.013166
2016-04-06    0.003175
2016-04-07    0.001354
Name: date_crawled, dtype: float64

Most of the dates for date_crawled lie in early march with hardly any dates coming after, according to the distribution shown.

In [121]:
autos["ad_created"].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
2015-12-30    0.000022
2016-01-03    0.000022
2016-01-07    0.000022
2016-01-10    0.000044
2016-01-13    0.000022
2016-01-14    0.000022
2016-01-16    0.000022
2016-01-22    0.000022
2016-01-27    0.000067
2016-01-29    0.000022
2016-02-01    0.000022
2016-02-02    0.000044
2016-02-05    0.000044
2016-02-07    0.000022
2016-02-08    0.000022
2016-02-09    0.000022
2016-02-11    0.000022
2016-02-12    0.000044
2016-02-14    0.000044
2016-02-16    0.000022
2016-02-17    0.000022
2016-02-18    0.000044
2016-02-19    0.000067
2016-02-20    0.000044
2016-02-21    0.000044
                ...   
2016-03-09    0.033037
2016-03-10    0.032482
2016-03-11    0.033370
2016-03-12    0.037123
2016-03-13    0.016985
2016-03-14    0.034880
2016-03-15    0.033770
2016-03-16    0.029840
2016-03-17    0.030817
2016-03-18    0.013477
2016-03-19    0.033659
2016-03-20    0.038210
2016-03-21 

Most of the value lie in the month of march and april for ad_created with negligible markings after the month of april

In [122]:
autos["last_seen"].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.001088
2016-03-06    0.004174
2016-03-07    0.005218
2016-03-08    0.007016
2016-03-09    0.009458
2016-03-10    0.010302
2016-03-11    0.012056
2016-03-12    0.023934
2016-03-13    0.008881
2016-03-14    0.012300
2016-03-15    0.015697
2016-03-16    0.016163
2016-03-17    0.027709
2016-03-18    0.007393
2016-03-19    0.015431
2016-03-20    0.020426
2016-03-21    0.020671
2016-03-22    0.021270
2016-03-23    0.018384
2016-03-24    0.019516
2016-03-25    0.018606
2016-03-26    0.016430
2016-03-27    0.015453
2016-03-28    0.020537
2016-03-29    0.021381
2016-03-30    0.024179
2016-03-31    0.023468
2016-04-01    0.022891
2016-04-02    0.024889
2016-04-03    0.024933
2016-04-04    0.024334
2016-04-05    0.126421
2016-04-06    0.225355
2016-04-07    0.134036
Name: last_seen, dtype: float64

Again most of the dates lie in april now with the other months having almost negliblie dates for the last_seen

In [123]:
autos["registration_year"].describe()

count    45040.000000
mean      2005.065542
std         89.707014
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Some unrealistic values such as 1000 and 9999 exist in the registration_year column, that have to be filtered out.

In [124]:
autos["registration_year"].value_counts().sort_index(ascending=True)

1000       1
1001       1
1910       2
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       1
1952       1
1953       1
1954       2
1955       1
1956       4
1957       2
1958       4
1959       6
1960      19
1961       6
1962       4
1963       7
1964      11
1965      17
1966      21
        ... 
2000    2714
2001    2519
2002    2419
2003    2657
2004    2682
2005    2864
2006    2665
2007    2269
2008    2204
2009    2076
2010    1581
2011    1614
2012    1302
2013     790
2014     650
2015     359
2016     944
2017    1296
2018     462
2019       1
2800       1
4100       1
4500       1
4800       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 93, dtype: int64

In [125]:
autos = autos[autos["registration_year"].between(1961,2016)].sort_values('registration_year',ascending=True)

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s. We have assumed years between 1961 and 2016 as the ones that are correct for our calculation as the number of cars beyond that are very few and negligible based on value_counts result.

In [126]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=False)

2016    0.021848
2015    0.008309
2014    0.015044
2013    0.018284
2012    0.030134
2011    0.037355
2010    0.036591
2009    0.048048
2008    0.051010
2007    0.052515
2006    0.061680
2005    0.066286
2004    0.062073
2003    0.061495
2002    0.055986
2001    0.058301
2000    0.062814
1999    0.059527
1998    0.046821
1997    0.034971
1996    0.024788
1995    0.019858
1994    0.011572
1993    0.007869
1992    0.007036
1991    0.006781
1990    0.006342
1989    0.003587
1988    0.002962
1987    0.001574
1986    0.001504
1985    0.001967
1984    0.001134
1983    0.001157
1982    0.000949
1981    0.000625
1980    0.001759
1979    0.000787
1978    0.000972
1977    0.000486
1976    0.000486
1975    0.000417
1974    0.000555
1973    0.000532
1972    0.000717
1971    0.000555
1970    0.000787
1969    0.000440
1968    0.000579
1967    0.000602
1966    0.000486
1965    0.000393
1964    0.000255
1963    0.000162
1962    0.000093
1961    0.000139
Name: registration_year, dtype: float64

Based on the above data we can observe that most registrations happened in between 1994 and 2016.

In [127]:
selected_brands = list(autos["brand"].value_counts(ascending=False, normalize=True).head(20).index)

In [128]:
selected_brands

['volkswagen',
 'bmw',
 'mercedes_benz',
 'opel',
 'audi',
 'ford',
 'renault',
 'peugeot',
 'fiat',
 'seat',
 'skoda',
 'smart',
 'nissan',
 'mazda',
 'citroen',
 'toyota',
 'hyundai',
 'volvo',
 'mini',
 'sonstige_autos']

The top 20 brands have been selected to be used as brands having a share more than 5% are very small in number.

In [129]:
brand_mean = dict()
for i in selected_brands:
    mean = autos[autos["brand"] == i]["price"].mean()
    brand_mean[i]=mean

In [130]:
brand_mean

{'audi': 9571.457397959184,
 'bmw': 8447.550823623946,
 'citroen': 3994.8823529411766,
 'fiat': 3243.8300589390965,
 'ford': 4197.397868561279,
 'hyundai': 5686.940909090909,
 'mazda': 4459.2692307692305,
 'mercedes_benz': 8568.987083616586,
 'mini': 10691.06157635468,
 'nissan': 5175.8969230769235,
 'opel': 3392.986414920562,
 'peugeot': 3360.9205974842766,
 'renault': 2772.7642105263158,
 'seat': 4810.883870967742,
 'skoda': 6564.499325236167,
 'smart': 3611.0442748091605,
 'sonstige_autos': 11504.05115089514,
 'toyota': 5266.97074010327,
 'volkswagen': 5780.4793505502885,
 'volvo': 5189.4236453201975}

sonstige_autos, mini, audi seem to have the highest mean prices among the dictionary of mean prices. 
fiat, opel, smart are the ones hacing the lowes mean prices among the dictionary items

In [131]:
comp_selected_brands = list(autos["brand"].value_counts(ascending=False, normalize=True).head(6).index)

In [132]:
new_brand_mean = dict()
for i in comp_selected_brands:
    mean = autos[autos["brand"] == i]["price"].mean()
    new_brand_mean[i]=mean

In [133]:
new_brand_mean

{'audi': 9571.457397959184,
 'bmw': 8447.550823623946,
 'ford': 4197.397868561279,
 'mercedes_benz': 8568.987083616586,
 'opel': 3392.986414920562,
 'volkswagen': 5780.4793505502885}

In [134]:
bmp_series = pd.Series(new_brand_mean)

In [135]:
new_mean_mileage = dict()
for i in comp_selected_brands:
    mean = autos[autos["brand"] == i]["odometer_km"].mean()
    new_mean_mileage[i]=mean

In [136]:
new_mean_mileage

{'audi': 128941.32653061225,
 'bmw': 132977.09923664122,
 'ford': 123664.29840142097,
 'mercedes_benz': 131312.03263086337,
 'opel': 128070.45820861156,
 'volkswagen': 128235.80690857579}

In [137]:
mmp_mileage = pd.Series(new_mean_mileage)

In [138]:
df_price_mileage = pd.DataFrame(bmp_series, columns=['mean_price'])

In [139]:
df_price_mileage["mean_mileage"] = mmp_mileage

In [140]:
df_price_mileage

mean_price   mean_mileage
audi           9571.457398  128941.326531
bmw            8447.550824  132977.099237
ford           4197.397869  123664.298401
mercedes_benz  8568.987084  131312.032631
opel           3392.986415  128070.458209
volkswagen     5780.479351  128235.806909

## Conclusion 

_The mileages of the brands are almost the same indiciating no distinct variation. This futher goes on to say that the more the opel, ford and to a certain extent volkswagen have run the lesser is their mean price_

## Further Analysis

### Damaged vs Not Damaged

In [141]:
autos['unrepaired_damage'].value_counts(dropna=False)

nein    32912
NaN      6739
ja       3556
Name: unrepaired_damage, dtype: int64

In [142]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace("nein","No")
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace("ja","Yes")

In [143]:
autos['unrepaired_damage'].value_counts(dropna=False)

No     32912
NaN     6739
Yes     3556
Name: unrepaired_damage, dtype: int64

In [144]:
damaged = autos[autos['unrepaired_damage'] == 'Yes']
not_damaged = autos[autos['unrepaired_damage'] == 'No']

In [145]:
damaged['price'].mean()

2755.9240719910013

In [146]:
not_damaged['price'].mean()

7101.729673067574

It is evident from the above values that the unrepaired damaged cars sell for a much lesser mean price which is 2755.92, compared to cars which are not damaged that sell at 7107.93.

### Most Popular Brand

In [147]:
autos['Manufacturer'] = autos['name'].str.split('_').str[0]

In [148]:
autos['Manufacturer'].value_counts().head(5)

Volkswagen    4925
BMW           4021
Mercedes      3788
Opel          3689
Audi          3432
Name: Manufacturer, dtype: int64

The above are the top 5 brands of cars that are most common brands in the car resale market.

### Effect of mileage on the price of Car

In [149]:
odometer_dict = {}

In [150]:
for i in list(autos['odometer_km'].value_counts().index):
    odometer_dict[i] = autos[autos['odometer_km'] == i]

In [151]:
average_price = {}
for i in list(autos['odometer_km'].value_counts().index):
    df = odometer_dict[i]
    average_price[i] = df['price'].mean()

In [152]:
average_price

{5000: 8407.308529945554,
 10000: 20065.05504587156,
 20000: 17101.9066091954,
 30000: 15885.32157394844,
 40000: 15019.718670076727,
 50000: 13588.431958762887,
 60000: 12216.389598540147,
 70000: 11123.188744588744,
 80000: 9779.593144560358,
 90000: 8703.25294482331,
 100000: 8218.59624174708,
 125000: 6427.662074903143,
 150000: 4082.1728638086906}

Excluding 5000kms we can see that the average price of the car decreases with increase in mileage, the price for 10,000kms is the highest and it continues to decrease. It is the lowest for 150,000kms.